File: src/bin/mt/apksign/data/FileDataSource.java
package bin.mt.apksign.data;

import bin.io.RandomAccessFile;

import java.io.EOFException;
import java.io.IOException;
import java.io.OutputStream;

public class FileDataSource implements DataSource {
    private RandomAccessFile randomAccessFile;
    private long start;
    private long size;
    private long pos;

    FileDataSource(RandomAccessFile randomAccessFile, long start, long size) {
        this.randomAccessFile = randomAccessFile;
        this.start = start;
        this.size = size;
    }


    @Override
    public long size() {
        return size;
    }

    @Override
    public long pos() {
        return pos;
    }

    @Override
    public void reset() {
        pos = 0;
    }

    @Override
    public void copyTo(OutputStream os, long length) throws IOException {
        if (length > remaining())
            throw new EOFException();
        byte[] buf = new byte[4096];
        int readLen;
        randomAccessFile.seek(start + pos);
        while (length > 0 && (readLen = randomAccessFile.read(buf, 0, (int) Math.min(length, buf.length))) != -1) {
            os.write(buf, 0, readLen);
            length -= readLen;
            pos += readLen;
        }
        if (length != 0)
            throw new IllegalStateException("Remaining length: " + length);
    }
}

File: Android/app/src/main/java/bin/mt/apksign/data/FileDataSource.kt
package bin.mt.apksign.data

import bin.io.RandomAccessFile
import java.io.EOFException
import java.io.IOException
import java.io.OutputStream
import kotlin.math.min

class FileDataSource internal constructor(
    private val randomAccessFile: RandomAccessFile,
    private val start: Long,
    private val size: Long
) : DataSource {
    private var pos: Long = 0
    override fun size(): Long {
        return size
    }

    override fun pos(): Long {
        return pos
    }

    override fun reset() {
        pos = 0
    }

    @Throws(IOException::class)
    override fun copyTo(os: OutputStream, length: Long) {
        var length = length
        if (length > remaining()) throw EOFException()
        val buf = ByteArray(4096)
        var readLen = 0
        randomAccessFile.seek(start + pos)
        while (length > 0 && randomAccessFile.read(buf, 0, min(length, buf.size.toLong()).toInt())
                .also { readLen = it } != -1
        ) {
            os.write(buf, 0, readLen)
            length -= readLen.toLong()
            pos += readLen.toLong()
        }
        check(length == 0L) { "Remaining length: $length" }
    }
}